tools/checkrelease.sh

```diff
@@ -19,114 +19,225 @@
-    echo "   in that directory (which should include nuttx and nuttx-apps"
-    echo "   sha512, asc, and tar.gz files), checks the release SHA512 and GPG "
-    echo "   signatures, and checks the unpacked directories for required "
-    echo "   files. Creates a temporary directory to do its work in."
+    echo "Usage: $0 [--verbose] [--url <URL-of-release-dir>] [--release <name-of-release] [--dir <path-to-directory>] [--tempdir <path-to-directory>]"
+    echo "   Given release full URL, release name, or a local directory, downloads or copies"
+    echo "   all files in that directory (which for a release should include nuttx and nuttx-apps, sha512, "
+    echo "   asc, and tar.gz files), checks the release SHA512 and GPG signatures, checks the unpacked "
+    echo "   directories for required files, and tries to build NuttX for sim:nsh."
+    echo
+    echo "   If tempdir is specified, it will be removed and recreated; if it is not specified, /tmp/nuttx-checkrelease"
+    echo "   is used."
+    echo
+    echo "   If --url or --release are given, nuttx and nuttx-apps tar.gz files are left in /tmp for the user to "
+    echo "   build for their platform."
+    echo
+    echo "   -V and --verbose are equivalent."
     echo
-    echo "   nuttx and nuttx-apps tar.gz files are left in the current"
-    echo "   directory."
-    echo 
     echo "Examples:"
     echo
-    echo "  $0 9.1.0-RC1"
-    echo "  $0 https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/"
+    echo "  $0 --release 9.1.0-RC1"
+    echo "  $0 --url https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/"
+    echo "  $0 --dir ./some-dir-that-has-nuttx-and-apps"
     echo
 }
 
-if [ "-h" == "$1" ]; then
-    usage
-    exit 0
-fi
+UNKNOWN=()
+URL=""
+DIRECTORY=""
+
+while [[ $# -gt 0 ]]
+do
+  key="$1"
+
+  case $key in
+      -U|--url)
+      shift
+      URL="$1"
+      ;;
+      -R|--release)
+      shift
+      RELEASE="$1"
+      URL="$BASE_URL/$RELEASE/"
+      ;;
+      -D|--dir)
+      shift
+      DIRECTORY="$(readlink -f $1)"
+      ;;
+      -T|--tempdir)
+      shift
+      TEMPDIR="$(readlink -f $1)"
+      ;;
+      -h|--help)
+      usage
+      exit 0
+      ;;
+      -V|--verbose)
+      VERBOSE=1
+      ;;
+      *)    # unknown option
+      UNKNOWN+=("$1") # save it in an array for later
+      ;;
+  esac
+  shift
+done
+set -- "${UNKNOWN[@]}" # restore unknown positional parameters
 
-if [ -z "$1" ]; then
-    usage
-    exit 0
+if [[ (-n "$UNKNOWN") ]]; then
+  usage
+  exit 1
 fi
 
-ARG=$1
-if [[ "$ARG" =~ ^"http".* ]]; then
-  URL="$1/"
-else
-  URL="$BASE_URL/$1/"
+if [[ (-z "$URL") && (-z "$DIRECTORY") ]]; then
+  usage
+  exit 1
 fi
 
 download_release
-check_nuttx 
-check_nuttx_apps 
+if [[ -n "$URL" ]]; then
+  # download and check
+  check_nuttx
+  check_nuttx_apps
+else
+  # check directories without downloading
+  check_required_files "nuttx"
+  check_required_files "apps"
+fi
+check_sim_nsh_build
```

Comment:
Do you try build sim:nsh with the download from website? I don't think the script can work as expect, because:
1.the unzip nuttx/apps is moved to /tmp
2.but we try to build /tmp/nuttx-release/nuttx

Reply:
@xiaoxiang781216 I'm not sure what you mean here. Yes, I'm testing the script several times with each modification. You can comment out the trap statement, add `--verbose`, run the script, and verify that `sim:nsh` is built in the correct directory. When I do that I can see it's using the correct temporary directory `/tmp/nuttx-checkrelease`.

Reply:
@xiaoxiang781216 Ok, you might be right, something is messed up with how I'm testing. Looking into it.

Reply:
@xiaoxiang781216 Ah, it's a difference in your invocation of wget vs. mine. When I change to yours, the dist.apache.org directory is not created. So I was testing correctly before, but when I changed to your wget just now, that broke the script.

Reply:
@xiaoxiang781216 Ok, pushed changes to make the `wget` invocation work the way you suggested. It requires a trailing slash at the end of every URL. If the user gives a URL I add the slash because extra slashes are ok.

My testing setup was ok— all the options (`--url`, `--release`, `--dir`) all work for me.